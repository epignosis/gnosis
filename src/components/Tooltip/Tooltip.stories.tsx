import React, { ReactNode } from "react";
import { Story } from "@storybook/react";
import { Button } from "../../";
import Tooltip, { TooltipProps } from "./Tooltip";

export default {
  component: Tooltip,
  title: "Components/Tooltip",
  argTypes: {
    placement: {
      control: {
        type: "select",
        options: ["top", "right", "left", "bottom"],
      },
    },
    as: {
      control: {
        type: "select",
        options: ["div", "span"],
      },
    },
  },
  decorators: [
    (story: () => ReactNode): JSX.Element => (
      <div style={{ marginTop: 150, display: "flex", justifyContent: "center" }}>{story()}</div>
    ),
  ],
};

export const Default: Story<TooltipProps> = (args) => (
  <div>
    <Tooltip {...args}>
      <Button>Hover me</Button>
    </Tooltip>
  </div>
);

Default.args = {
  placement: "top",
  content: "This is a tooltip",
  as: "div",
};
