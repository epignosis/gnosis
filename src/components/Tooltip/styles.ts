import { css, Theme, SerializedStyles } from "@emotion/react";

export const tooltipContainer = ({ typeScaleSizes, tooltip }: Theme): SerializedStyles => {
  return css`
    background: ${tooltip.background};
    border-color: ${tooltip.background};
    color: ${tooltip.color};
    padding: 0.5rem;
    font-size: ${typeScaleSizes["2xs"]};

    &.tooltip[data-placement^="bottom"] > #arrow {
      top: -1px;
    }

    &.tooltip[data-placement^="left"] > #arrow {
      right: -4px;
    }

    &.tooltip[data-placement^="right"] > #arrow {
      left: -4px;
    }

    #arrow,
    #arrow::before {
      position: absolute;
      width: 1rem;
      height: 0.5rem;
      z-index: -1;
    }

    #arrow::before {
      content: "";
      transform: rotate(45deg);
      background: ${tooltip.background};
    }
  `;
};
