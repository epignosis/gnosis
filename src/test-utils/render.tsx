import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
