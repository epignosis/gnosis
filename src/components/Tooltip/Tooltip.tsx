import React, { FC, ReactNode } from "react";
import Tippy, { TippyProps } from "@tippyjs/react/headless";
import { tooltipContainer } from "./styles";

export type TooltipProps = TippyProps & {
  as?: "div" | "span";
  content: TippyProps["content"];
};

const Tooltip: FC<TooltipProps> = ({
  children,
  content,
  as = "div",
  placement = "top",
  ...rest
}) => {
  const Tag = as;

  return (
    <Tippy
      interactive
      placement={placement}
      render={(attrs): ReactNode => (
        <div className="tooltip" css={tooltipContainer} {...attrs}>
          {content}
          <div id="arrow" data-testid="tooltip-arrow" data-popper-arrow />
        </div>
      )}
      {...rest}
    >
      <Tag>{children}</Tag>
    </Tippy>
  );
};

export default Tooltip;
