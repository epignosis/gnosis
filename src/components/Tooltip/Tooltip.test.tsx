import React from "react";
import faker from "faker";
import userEvent from "@testing-library/user-event";
import Tooltip from "./Tooltip";
import { screen, render, act } from "@test-utils/render";

describe("<Tooltip/>", () => {
  it("renders correctly", async () => {
    const btnTxt = faker.lorem.word();
    const contentTxt = faker.lorem.word();

    render(
      <Tooltip content={contentTxt}>
        <button>{btnTxt}</button>
      </Tooltip>,
    );

    await act(async () => {
      userEvent.hover(screen.getByText(btnTxt));
    });

    const buttonText = screen.getByText(btnTxt);
    const content = screen.getByText(contentTxt);
    const arrow = screen.getByTestId("tooltip-arrow");

    expect(buttonText).toHaveTextContent(btnTxt);
    expect(content).toHaveTextContent(contentTxt);
    expect(arrow).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Tooltip content="Test content">
        <button>test hover</button>
      </Tooltip>,
    );

    expect(container).toMatchSnapshot();
  });
});
