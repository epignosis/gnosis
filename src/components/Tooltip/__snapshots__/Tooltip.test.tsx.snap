// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip/> matches snapshot 1`] = `
<div>
  <div
    aria-expanded="false"
  >
    <button>
      test hover
    </button>
  </div>
</div>
`;
